User guide for persistent GCS w/ redis and kuberay (#49887)
ray-project/ray: doc/source/cluster/kubernetes/user-guides/kuberay-gcs-persistent-ft.md

```

On AWS, you must [Create a storage
class](https://docs.aws.amazon.com/eks/latest/userguide/create-storage-class.html)
yourself as well.

## Tuning backups

Redis supports database dumps at set intervals, which is good for fast recovery
and high performance during normal operation.

Redis also supports journaling at frequent intervals (or continuously), which
can provide stronger durabililty at the cost of more disk writes (i.e., slower
performance).

A good starting point for backups is to enable both, which can be done like so:

```
# Dump a backup every 60s, if there are 1000 writes since the prev. backup.
save 60 1000
dbfilename dump.rdb

# Enable the append-only log file.
appendonly yes
appendfilename "appendonly.aof"

```

In this recommended configuration, full backups are created every 60s while the
append-only log is updated every second, which is a reasonable balance for disk
space, latency, and data safety.

There are more options to configure the AOF, defaults shown here:

```
# Sync the log to disk every second.
# Alternatives are "no" and "always" (every write).
appendfsync everysec
auto-aof-rewrite-percentage 100
auto-aof-rewrite-min-size 64mb
```

> You can view the full reference for 5.0.8

Reviewer: ```suggestion
You can view the full reference
```

Author: Done.

[here](https://raw.githubusercontent.com/redis/redis/refs/tags/7.4.0/redis.conf).

If your job is generally idempotent and can resume from several minutes of state
loss, you may prefer to disable the append-only log.

If you prefer your job to lose as little state as possible, then you may prefer
to set `appendfsync` to `always`, such that all write are stored immediately.

> ## Putting it together

Reviewer 2: Would it be useful to show some steps for how a user could validate that persistence is working? Something like:
1) create an actor
2) restart the redis server
3) restart the head pod
4) confirm actor comes back

Author: Added a section at the bottom w/ verifying job metadata is persisted.

# Enable the append-only log file.
appendonly yes
appendfilename "appendonly.aof"

```

In this recommended configuration, full backups are created every 60s while the
append-only log is updated every second, which is a reasonable balance for disk
space, latency, and data safety.

There are more options to configure the AOF, defaults shown here:

```
# Sync the log to disk every second.
# Alternatives are "no" and "always" (every write).
appendfsync everysec
auto-aof-rewrite-percentage 100
auto-aof-rewrite-min-size 64mb
```

You can view the full reference
[here](https://raw.githubusercontent.com/redis/redis/refs/tags/7.4.0/redis.conf).

If your job is generally idempotent and can resume from several minutes of state
loss, you may prefer to disable the append-only log.

If you prefer your job to lose as little state as possible, then you may prefer
to set `appendfsync` to `always`, such that all write are stored immediately.

## Putting it together

Edit [the full
YAML](https://github.com/ray-project/kuberay/blob/master/config/samples/ray-cluster.persistent-redis.yaml)
to your satisfaction and apply it:

```
kubectl apply -f config/samples/ray-cluster.persistent-redis.yaml
```

> Verify that a disk has been provisioned and redis is running:

Reviewer 3: ```suggestion
Verify that Kubernetes provisioned a disk and Redis is running:
```
is kubernetes provisioning the disk?

Author: I think it's fair to say so

```
kubectl get persistentvolumes
kubectl get pods
# Should see redis-0 running.
```

> After running a job with some state in GCS, you will be able to delete the ray
> head pod as well as the redis pod without data loss.

Reviewer 3: ```suggestion
After running a job with some state in GCS, you can delete the ray
head pod as well as the redis pod without data loss.
```

## Verifying

> Forward connections to the ray cluster you just created with the Ray Kubectl

Reviewer 2: hyper link Ray Kubectl plugin to https://docs.ray.io/en/latest/cluster/kubernetes/user-guides/kubectl-plugin.html

plugin:

```
$ kubectl ray session raycluster-external-redis
```

Then submit any Ray job of your choosing and let it run. When finished, delete
all your pods:

```
$ kubectl delete pods --all
```

> Wait for the ray head to be provisioned again and enter ready state. Then
> restart your port forwarding and view the ray dashboard. You should find the
> job's metadata has been persisted, despite the loss of the ray head as well as
> the redis replica.

Reviewer 3: ```suggestion
Wait for Kubernetes to provision the Ray head and enter a ready state. Then
restart your port forwarding and view the Ray dashboard. You should find that Kubernetes has persisted your job's metadata, despite the loss of the ray head as well as
the Redis replica.
```
is kubernetes doing the provisioning and entering the ready state? is it persisting the job's metadata?

Author: Hm, the metadata is stored on disk and restored by redis and then the ray head.
